A graphical user interface for editing and designing.

Wang-yuanlu opened the pull request:
Hi, Mr Tan. I made some adjustments to the GUI interface and added a black border to each panel to make them look better. Here are the next GUI optimization goals I think I need to do:
1. The tile size in the cgrapanel and mappingpanel can be further adjusted. A 4*4 size model will have some gaps, a 5*5 model size is appropriate, while a 6*6 size is a bit too large.
2. The size of parampanel and testpanel can be further adjusted to make the whole GUI interface look better.
3. The size of DFG can be further adjusted to make the kernelpanel look better.
![GUI 4×4](https://github.com/tancheng/CGRA-Flow/assets/80200672/5cf44251-e424-42ca-8dcc-5d92c2f69fd6)
![GUI 5×5](https://github.com/tancheng/CGRA-Flow/assets/80200672/4eea08c6-db67-4479-a71f-b88b1c590226)
![GUI 6×6](https://github.com/tancheng/CGRA-Flow/assets/80200672/3743d21c-0181-40a4-9612-bc257ce0b5ce)

tancheng reviewed (commented):
For the solid borders:

Can you also try to add border into the components inside the `kernel panel` with a screenshot? And would it look better if each button also get the solid border?

For your TODOs:
> The tile size in the cgrapanel and mappingpanel can be further adjusted. A 44 size model will have some gaps, a 55 model size is appropriate, while a 6*6 size is a bit too large.

I think it looks fine for the `5x5` and `6x6`, as we have the horizontal and vertical scrollbars. For the `4x4`, it is better to be adjusted.

I think the issue is the the CGRA panel is not a square and you are determining the size of the tile based on the panel width. Instead, you can determine the tile size based on the `max(panel_height, panel_width)`, and try best to make the `CGRA panel row` and the `kernel panel row` have the same height when the window is popped up.

> The size of parampanel and testpanel can be further adjusted to make the whole GUI interface look better.

Yes, thanks for taking this into consideration, which can be done in the next PR (or this one, you decide).

> The size of DFG can be further adjusted to make the kernelpanel look better.

It looks fine to me now, the only issue is the DFG figure has white background (rather than transparent), and even it is transparent, it cannot fit into the panel. So the solution maybe try to "stretch" the `DFG.png` to fit it into the panel? I believe the tkinter should have such library/interface for that.

You can decide what will be done in each PR, just let me whether existing PR is ready for merge.

Wang-yuanlu commented on launchUI.py:
```diff
@@ -1421,8 +1421,8 @@ def create_cgra_pannel(master, rows, columns):
     TILE_HEIGHT = (GRID_HEIGHT + LINK_LENGTH) / ROWS - LINK_LENGTH
-
-    cgraPannel = tkinter.LabelFrame(master, text='CGRA', bd=BORDER, relief='groove')
+    #Use solid black board to let the pannel look better
```
Okay I will do it.

Wang-yuanlu commented on launchUI.py:
```diff
@@ -1692,13 +1696,13 @@ def create_param_pannel(master):
     dataPannel =tkinter.LabelFrame(master)
     dataPannel.grid(row=0, column=2, rowspan=1, columnspan=1, sticky="nsew")
-    dataPannel.grid_rowconfigure(0, weight=1)
+    # Increase the size of the 'Sverilog' panel
```
Okay I will do it.

Wang-yuanlu commented on launchUI.py:
```diff
@@ -1738,53 +1742,53 @@ def create_test_pannel(master):
 
-    synthesisTimeEntry.grid(row=1, column=0, pady=10)
-    reportTimecostLabel.grid(row=1, column=1, pady=10)
+    synthesisTimeEntry.grid(row=1, column=1, pady=10)
```
Certainly, I realized that 'time entry' used elements from the 'ttk' library, and after unifying them as elements from the 'tkinter' library, they align correctly."

Wang-yuanlu commented on launchUI.py:
```diff
@@ -1598,16 +1602,16 @@ def create_param_pannel(master):
     widgets["dataMemEntry"] = dataMemEntry   
     updateButton = tkinter.Button(paramPannel, text = " Reset ", relief='raised', command = partial(clickReset, master))
-    updateButton.grid(row=2, column=2)
+    updateButton.grid(row=2, column=2, columnspan=2, sticky="nsew")
```
Certainly, I adjusted the relative positions of the buttons to make them align.

Wang-yuanlu commented:
Hi, Mr Tan. Following your modification suggestions, I made further adjustments and improvements to the GUI interface. I added a black solid board to each component and adjusted the relative sizes of 'parampanel' and 'testpanel', which also made the relative area of 'cgrapanel' more appropriate. Additionally, I optimized the process of generating DFG, allowing the generated DFG graph to be transparently embedded in the corresponding panel.
Here are the next GUI optimization goals I think I need to do:
1. Try to determine the tile size based on the max(panel_height, panel_width).
2. Try to make the CGRA panel row and the kernel panel row have the same height when the window is popped up. However, I believe that this issue is determined by the tkinter library itself. In order to minimize the impact of this issue and ensure that 'cgrapanel' occupies an appropriate area in terms of space, I adjusted the GUI layout to its current style. I believe this can have a certain effect.
![GUI 4×4_2](https://github.com/tancheng/CGRA-Flow/assets/80200672/f33d7e83-2e5c-45cb-97b0-75556760323b)

tancheng commented on launchUI.py:
```diff
@@ -1225,10 +1220,27 @@ def clickShowDFG():
     convertProc = subprocess.Popen([convertCommand, "-u"], stdout=subprocess.PIPE, stderr=subprocess.PIPE, shell=True)
     (out, err) = convertProc.communicate()
 
+    # Get the size of the whole window
```
Lets use `Gets` instead of `Get`.

Wang-yuanlu replied:
Okay I will do it.

tancheng commented on launchUI.py:
```diff
@@ -1225,10 +1220,27 @@ def clickShowDFG():
+    new_data = []
+    for item in datas:
+        if item[0] == 255 and item[1] == 255 and item[2] == 255:
+            # Make the image transparent where it is white
```
nit: `Makes`.

Wang-yuanlu replied:
Okay I will do it.

tancheng commented on launchUI.py:
```diff
@@ -1311,7 +1323,7 @@ def drawSchedule():
     for ii in range(mappingII):
         # draw data memory
-        spmLabel = tkinter.Label(canvas, text="Data\nSPM", fg='black', bg='gray', relief='raised', bd=BORDER)
+        spmLabel = tkinter.Label(canvas, text="Data\nSPM", fg='black', bg='gray', relief='raised', bd=BORDER, highlightbackground="black", highlightthickness=2)
```
Would it look better if we make the border a little thinner? Say `thickness = 1`? WDYT?

tancheng replied:
And I think the `thickness` could also be a global constant variable, right?

Wang-yuanlu replied:
Certainly, I set the thickness as a global variable and I set the thickness to 1. It seems that let thickness = 1 will make the graphical user interface more clear and concise.
![Uploading GUI thickness=1.png…]()

tancheng reviewed (commented):
Hi Yuanlu, thanks for the quick fix.

 - You can actually reply reviewer's comment at the corresponding place after you made (commit) the change, so it would be more clear to the reviewer what is changed.
 - Do you think it would look better if we also add border to the `label`s?
 - Can we avoid adding border to the `panel`? The panel itself has some border with name embedded on top, adding border to panel makes it look weird.
 - How did you make the `DFG.png` be transparent? It is done outside the GUI using other tools? If yes, the issue would still be there, e.g., when you enlarge/shrink the window, the showing `DFG` won't be scaled up/down.

Wang-yuanlu commented:
Hi, Mr Tan. Following your modification suggestions, I made further adjustments and improvements to the GUI interface.
I change the thickness of the solid borad to make the GUI more clear and concise. I used functions from PIL library which can be used to create transparent backgrounds or overlay the image on top of other content and then optimizing the DFG.
![GUI thickness=1](https://github.com/tancheng/CGRA-Flow/assets/80200672/a725f8ff-fa89-4662-b850-721799827ed1)

Wang-yuanlu commented:
Hi, Mr Tan. Following your modification suggestions, I made further adjustments and improvements to the GUI interface.
I change the thickness of the solid borad to make the GUI more clear and concise. I used functions from PIL library which can be used to create transparent backgrounds or overlay the image on top of other content and then optimizing the DFG.
The results are as follows. Have a nice day!
![GUI thickness=1](https://github.com/tancheng/CGRA-Flow/assets/80200672/48030f36-7675-4572-8358-dd8b5a4d2cd6)

tancheng commented on launchUI.py:
```diff
@@ -30,6 +30,7 @@
 MEM_WIDTH = 50
 CONFIG_MEM_SIZE = 8
 DATA_MEM_SIZE = 4
+thickness = 1
```
Plz rename this as `HIGHLIGHT_THICKNESS`.

Wang-yuanlu replied:
okay I will do it.

tancheng commented:
Hi Yuanlu, thanks for the update.
 - Is it possible to add a tiny interval between labels/fields. They are too closed to each other. Say,
   - `Time cost:` and its corresponding field in "report area/power".
   - `Time cost:`'s field is too closed to the boundary of the entire panel.
   - Also the `Application:`'s field and the `Compile app`.
 - Now the `DFG.png` can be stretched, right? Can you plz attach an enlarged window to show case the stretched `DFG.png`? I am thinking maybe the white background is more suitable than the transparent version.

Wang-yuanlu commented:
Hi Mr.Tan. I have further optimized the positioning of the components in the param module, and adjusted the spacing between the components in the GUI as well as the spacing between the components and the boundaries. Regarding the DFG diagram part, I have adjusted it back to a white background.

tancheng commented:
Hi Yuanlu, thanks for the update, can u also share the screenshot?

Wang-yuanlu commented:
> Hi Yuanlu, thanks for the update, can u also share the screenshot?

Okay. Here is the screenshot:
![GUI now](https://github.com/user-attachments/assets/aa698b77-7665-4230-9803-f93748d3de7e)

tancheng commented:
Cool, can u stretch out the window and paste the screenshot again? I wanna take a look at the stretched out DFG graph. Thanks!

Wang-yuanlu commented:
> Cool, can u stretch out the window and paste the screenshot again? I wanna take a look at the stretched out DFG graph. Thanks!

Okay. 
![GUI now 2 0](https://github.com/user-attachments/assets/f94b4950-14a8-4767-850b-3471e88eee98)
However, the stretching effect of the DFG diagram is not very pronounced.

tancheng commented:
Do u think it is possible to make the `DFG.png` automatically align with the panel size (we can sacrifice the resolution)?

Wang-yuanlu commented:
> Do u think it is possible to make the `DFG.png` automatically align with the panel size (we can sacrifice the resolution)?

I think that I am currently unable to adjust the adaptation of the DFG graph. It is embedded as a fixed-size component within the kernel panel and I have tried to add scrollbars for it, but the corresponding panel structure cannot accommodate the scrollbars.

tancheng commented:
No problem. Thanks! Let me merge the PR.
